ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -251,20 +258,43 @@ protected Node visitJoin(final Join node, final AnalysisContext context) {
     return null;
   }
 
-  private String fetchKeyFieldName(Expression expression) {
+  private Pair<String, String> fetchKeyFieldName(ComparisonExpression comparisonExpression, String sourceAlias, Schema
+      sourceSchema) {
+    Pair<String, String> keyInfo = fetchKeyFieldNameFromExpr(comparisonExpression.getLeft(),
+                                                      sourceAlias, sourceSchema);
+    if (keyInfo == null) {
+      keyInfo = fetchKeyFieldNameFromExpr(comparisonExpression.getRight(), sourceAlias, sourceSchema);
+    }
+    if (keyInfo == null) {
+      throw new KsqlException(String.format("%s : Invalid join criteria %s ", comparisonExpression
+          .getLocation().get().toString(), comparisonExpression));
+    }
+    return keyInfo;
+  }
+
+  private Pair<String, String> fetchKeyFieldNameFromExpr(Expression expression, String sourceAlias,
+                                                  Schema sourceSchema) {
     if (expression instanceof DereferenceExpression) {
       DereferenceExpression
-          leftDereferenceExpression =
+          dereferenceExpression =
           (DereferenceExpression) expression;
-      return leftDereferenceExpression.getFieldName();
+      String sourceAliasVal = dereferenceExpression.getBase().toString();
```

Comment:
I find this confusing. Can we at least add a method to `DereferenceExpression` called `getSourceAlias()`?
Ideally we can come up with an interface that both `DereferenceExpression` and `QualifiedNameReference` implement that enable us to get the `keyFieldName`. We can then get rid of the `instanceof` checks here and delegate to the impl

Reply:
This would be a bit difficult since these are two different nodes in our AST and SourceAlias is not always the lefthand side for these expressions. The interpretation of these expression also depends on the context.